server/src/main/java/org/candlepin/model/Product.java

```diff
@@ -1205,9 +1206,16 @@ public boolean equals(Object obj) {
                 .append(this.name, that.name)
                 .append(this.multiplier, that.multiplier)
                 .append(this.attributes, that.attributes)
-                .append(this.derivedProduct, that.derivedProduct)
                 .isEquals();
 
+            // When checking children versioned entities, we're only interested in verifying that the
+            // reference is the same, otherwise we could run into some weird linkage problems (also it's
+            // much faster)
+            Comparator<Product> productUuidComparator = (p1, p2) ->
+                Objects.equals(p1 != null ? p1.getUuid() : null, p2 != null ? p2.getUuid() : null) ? 0 : 1;
```

Comment:
3 ternary operators on a single line are a bit scary. I would suggest taking advantage of Comparator.comparing to compose the comparator.

Something like this.

```
Comparator.comparing(p -> p != null ? p.getUuid() : null, Comparator.nullsLast(Comparator.naturalOrder()))
```

Reply:
An alternate approach here could be to update the ```Util.collectionsAreEqual``` to use a ```BiFunction``` or a custom functional interface instead of a Comparator. Those utilities are fairly old now, and were written before functional interfaces were available to us, and without the knowledge of the BiFunction interface.  
At that point we get rid of one of the ternary operations that normalizes the output for what a Comparator expects.

Might be out of scope for this PR, however.

Reply:
Changed to use ```comparing```.